text/0000-new-test-waiters.md

```diff
@@ -0,0 +1,248 @@
+
+## New Test Waiters Design
+
+The new test waiters addon is built using low-level primitives that are complimented with some convenience utilities.
+
+### `IWaiter`
+
+At its core, the addon uses an `IWaiter` interface defined as follows:
+
+```ts
+export type WaiterName = string;
+export type Token = unknown;
+
+export interface IWaiter {
+  name: WaiterName;
+  waitUntil(): boolean;
+  debugInfo(): ITestWaiterDebugInfo[];
+}
+```
+
+- `name`: The name of the test waiter, which is used to help identify it in test isolation validation output.
+- `waitUntil`: Used to determine if the waiter system should still wait for async
+operations to complete. The `waitUntil` method will return `true` to signal completion.
+- `debugInfo`: Returns the `debugInfo` for each item tracking async operations in a waiter. The `debugInfo` for each waiter item is ultimately used in `@ember/test-helpers`' `getSettledState` function, which is used for test isolation validation output.
+
+This allows for maximum flexibility when creating your own waiter implementations.
+
+### `ITestWaiter`
+
+The `IWaiter` interface is built upon to create a more specific interface for a test waiter, `ITestWaiter`:
+
+```ts
+export interface ITestWaiter<T = Token> extends IWaiter {
+  beginAsync(token?: T, label?: string): T;
+  endAsync(token: T): void;
+  reset(): void;
+}
+```
+
+- `beginAsync`: Should be used to signal the beginning of an async operation that
+is to be waited for. Invocation of this method should be paired with a subsequent
+`endAsync` call to indicate to the waiter system that the async operation is completed.
+- `endAsync`: Should be used to signal the end of an async operation. Invocation of this
+method should be paired with a preceding `beginAsync` call, which would indicate the
+beginning of an async operation.
+- `reset`: Resets the waiter state, clearing items tracking async operations in this waiter.
+
+This interface is used for the concrete `TestWaiter` type. This type forms the basis for the addon, and will likely satisfy the majority of use cases.
+
+The most common practice is to import and invoke the `buildWaiter` function to create a new test waiter. The recommendation is to do so at the module level, which allows a single waiter to be created per type (this should likely be enforced via a lint rule added to `eslint-plugin-ember`). A single waiter is then usable across multiple instances.
+
+```ts
+function buildWaiter(name: string): ITestWaiter
+```
+
+In anything but a production build, this function will return a `TestWaiter` instance. When in production mode, we make this instance _inert_ and essentially no cost to invoke. Since test waiters are intended to be called from application or addon code, but are only required to be _active_ when in tests, this process of making the instance _inert_ is important. Even though code is still invoked, this has a negligible impact on performance.
```

Comment:
How do we feel about the plausibility of a system which would make this configurable. Basically users who want to optimize for test-ability of production builds can leave the inert code, and users who want to optimize for every byte can strip the inert code out?

Or is this already designed in such a way that it would be trivial to author an addon which stripped the inert code from production builds, allowing byte-watchers to opt-in to that size optimization? (I think it would be easy, but I'd like to hear from the authors)

Reply:
I think there are two parts to this:

1. ensuring that the addon's code itself is not shipped outside of testing
2. authoring a babel plugin that allows the statements in app and addon code to be removed

Both are doable and should be investigated during implementation (though I don't think it's particularly important to happen _before_ we land the RFC).

Reply:
These are good callouts, @mixonic. Early on @rwjblue and I discussed the possibility of stripping code from prod builds that weren't destined to be tested. I think as Rob describes above, we could absolutely do this. 

We also need to add a separate flag other than `DEBUG`, such as `FOR_TESTING` (don't worry about the name; naming is hard yada yada), but that will allow us to do prod builds that will allow for testing. Obviously this is a broader change that impacts a number of things, but is something we've also discussed.